ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/BigLedgerPeers.hs

```diff
@@ -0,0 +1,223 @@
+                                             , IsLedgerPeer
+                                             ))
+                                      newPeers)
+                         (knownPeers st)
+
+            -- We got a successful response to our request, but if we're still
+            -- below target we're going to want to try again at some point.
+            -- If we made progress towards our target then we will retry at the
+            -- suggested ttl. But if we did not make progress then we want to
+            -- follow an exponential backoff strategy. The max retry time of 2^8
+            -- seconds is just over four minutes.
+            bigLedgerPeerBackoffs' :: Int
+            bigLedgerPeerBackoffs'
+              | Set.null newPeers = (bigLedgerPeerBackoffs st `max` 0) + 1
+              | otherwise         = 0
+
+            bigLedgerPeerRetryDiffTime :: DiffTime
+            bigLedgerPeerRetryDiffTime
+              | bigLedgerPeerBackoffs' == 0
+                          = ttl
+              | otherwise = 2^(bigLedgerPeerBackoffs' `min` 8)
+
+            bigLedgerPeerRetryTime :: Time
+            bigLedgerPeerRetryTime = bigLedgerPeerRetryDiffTime `addTime` now
+
+         in Decision {
+               decisionTrace = [TraceBigLedgerPeersResults
+                                 newPeers
+                                 bigLedgerPeerBackoffs'
+                                 bigLedgerPeerRetryDiffTime],
+               decisionState = st {
+                                 bigLedgerPeers              = bigLedgerPeers',
+                                 knownPeers                  = knownPeers',
+                                 bigLedgerPeerBackoffs       = bigLedgerPeerBackoffs',
+                                 bigLedgerPeerRetryTime      = bigLedgerPeerRetryTime,
+                                 inProgressBigLedgerPeersReq = False
+                               },
+               decisionJobs  = []
+             }
+
+
+aboveTarget :: forall m peeraddr peerconn.
+               (Alternative (STM m), MonadSTM m, Ord peeraddr)
+            => MkGuardedDecision peeraddr peerconn m
+aboveTarget PeerSelectionPolicy {policyPickColdPeersToForget}
+            st@PeerSelectionState {
+                 bigLedgerPeers,
+                 knownPeers,
+                 establishedPeers,
+                 inProgressPromoteCold,
+                 targets = PeerSelectionTargets {
+                             targetNumberOfKnownBigLedgerPeers
+                           }
+               }
+    -- Are we above the target for number of known big ledger peers
+    | numKnownBigLedgerPeers > targetNumberOfKnownBigLedgerPeers
+
+    -- Are there any cold big ledger peers we could pick to forget?
+    , numKnownBigLedgerPeers > numEstablishedBigLedgerPeers
+
+    , let numPeersCanForget = numKnownBigLedgerPeers
+                            - targetNumberOfKnownBigLedgerPeers
```

Comment:
Let's move `numPeersCanForget` to the `Guarded` constructor, to limit its scope.